change background to true

sotoiwa opened the pull request:
Since this policy allows for background scanning, I changed the background attribute true.

Closes #97

chipzoller commented:
If you wanted to take this up, it would be great to lump all these changes into a single PR rather than just a single policy update.

sotoiwa commented:
@chipzoller Would you please review my PR? The changes I have made are as follows:

- Specified validationFailureAction and background attribute in all policies including mutate and generate.
- `validationFailureAction` is now `audit` for all policies  including mutate and generate.
- `background` is now `false` for the mutate and generate policies.
- For the validate policy, `background` is `true` if possible.
  - I used the `kyverno verify` command to check if the `background` can be set to `true`. 
- Wrote `validationFailureAction` and `background` attributes in this order.
- Removed unnecessary trailing spaces and newlines.
- Some of the yaml was indented with a single space, so I fixed it to indent with two spaces.

chipzoller commented:
Because `validationFailureAction` is only applicable to `validate` rules, I would recommend eliminating that line from `mutate` and `generate` policies because it's irrelevant, ignored by Kyverno, and bloats (if only slightly) sample policies.

chipzoller commented on other/verify_image.yaml:
```diff
@@ -16,7 +16,7 @@ metadata:
       how to configure a similar rule and will require replacing with your image(s) and keys.
 spec:
-  validationFailureAction: enforce
+  validationFailureAction: audit
```
This one should remain in `enforce` mode because it's pointless to have it in `audit` mode.

sotoiwa replied:
It seemed to me that other `background: false` validate policies (e.g. block-pod-exec-by-pod-label.yaml) are also useless in `audit` mode. I don't understand why only this `verify_image.yaml` and the next `block_updates_deletes.yaml` should be in `enforce` mode.

chipzoller replied:
The samples that block operations are not really valuable in background mode, so setting those to `false` makes sense. And same thing with these being in `audit` mode. If the whole point of the policy is to block updates and deletes, putting in audit does nothing. That's the exception. With the verify image policy, same sort of deal except that policy is really only an illustration, and no user is actually going to use that policy as-is without modification.

sotoiwa replied:
Does the verifyImages rule allow background scan? I couldn't find any mention of it in the documentation. If background scan is possible, I thought it would be better to specify `background: true`.

chipzoller replied:
Kyverno v1.4.2 allows one to create this policy with `background` in either mode. @JimBugwadia what is supported here officially? I'm guessing `background: true` doesn't actually do anything. Can you confirm? If that ends up being the case, shouldn't we prohibit such policies from being created/updated?

sotoiwa replied:
I have changed the `validationFailureAction` back to `enforce`, and the `background` is `false` as in the original.

chipzoller commented on other/block_updates_deletes.yaml:
```diff
@@ -13,7 +13,7 @@ metadata:
       a cluster-admin.
 spec:
-  validationFailureAction: enforce
+  validationFailureAction: audit
```
This one should remain in `enforce` mode because it's pointless to have it in `audit` mode.

sotoiwa replied:
It seemed to me that other `background: false` validate policies (e.g. block-pod-exec-by-pod-label.yaml) are also useless in `audit` mode.  What do you think?

chipzoller replied:
The exec policies should probably be in `enforce` mode, but having those in background mode would help users get a sense if a Pod was exec'd into to assess the value of changing such a policy to `enforce`.

sotoiwa replied:
The block-pod-exec-* policy uses `request.operation`, so it can't be `background: true`. Should I still leave it as `audit` in this case?

block_updates_deletes.yaml and block-pod-exec-* are very similar rules, so it seems strange that these rules have different values for `validationFailureAction`.

chipzoller replied:
Any of the five block policies which use the CONNECT feature probably need to have `background: false`.

sotoiwa replied:
Sorry, my point was not clear.

Those pod-exec-* policies are already `background: false` and I have not changed `background`.

Those pod-exec-* policies have `validationFailureAction: audit` which I have not changed also. On the other hand, block_updates_deletes.yaml has `validationFailureAction: enforce`, and I changed it to `validationFailureAction: audit` because I think pod-exec-* and block_updates are same kind of rule.

If we leave `validationFailureAction: enforce` of block_updates as it is, then I think other policies that require `background: false` because of `requests.operation`, such as block-pod-exec-*, should also have `validationFailureAction: enforce`.

chipzoller replied:
Ok, let's standardize on policies that validate for non-CREATE operations (UPDATE, CONNECT, DELETE) should be set `validationFailureAction: enforce`. Background mode should be set accordingly to `false`.

sotoiwa replied:
Thank you for the comment. I changed policies that validate for non-CREATE operations (UPDATE, CONNECT, DELETE) to have validationFailureAction: enforce.

sotoiwa commented:
> Because validationFailureAction is only applicable to validate rules, I would recommend eliminating that line from mutate and generate policies because it's irrelevant, ignored by Kyverno, and bloats (if only slightly) sample policies.

I agree with you. I will remove the `validationFailureAction` attribute from all mutate and generate policies.

I think the `background` attribute is also only applicable to validate rules and ignored by kyverno in mutate and generate polices, but I'll leave `background: false` listed to remind people that it won't be background scanned.

chipzoller commented:
> 
> 
> > Because validationFailureAction is only applicable to validate rules, I would recommend eliminating that line from mutate and generate policies because it's irrelevant, ignored by Kyverno, and bloats (if only slightly) sample policies.
> 
> I agree with you. I will remove the `validationFailureAction` attribute from all mutate and generate policies.
> 
> I think the `background` attribute is also only applicable to validate rules and ignored by kyverno in mutate and generate polices, but I'll leave `background: false` listed to remind people that it won't be background scanned.

I'd suggestion eliminating it entirely. Otherwise, users may be under the false impression that if they change `false` to `true` on such policies that they'll work retroactively, and that's not how Kyverno works (at present, anyway).

sotoiwa commented:
> I'd suggestion eliminating it entirely. Otherwise, users may be under the false impression that if they change false to true on such policies that they'll work retroactively, and that's not how Kyverno works (at present, anyway).

Thank you for the suggestion. I eliminated both `validationFailureAction` and `background` attribute from all mutate and generate policies.

chipzoller commented:
Once all questions/conversations are resolved, indicate when this PR is ready for final review.

chipzoller commented:
Is this ready for final review, @sotoiwa ?

sotoiwa commented:
@chipzoller Yes, it's ready for review. Thanks.

sotoiwa commented:
@chipzoller Would you please review the PR?

chipzoller commented:
Apologies. Looks like we may need to hold this because the test file result will be changing with a commit that was made recently. Rather than merge this PR and go back to update all of them again, we might just wait until 1.4.3 is cut and update the files.

chipzoller commented:
@vyankyGH can we accept this in 1.4.3 now?

chipzoller commented:
@viveksahu26 see above

vyankyGH commented:
> @vyankyGH can we accept this in 1.4.3 now?

Let me go through all the changes. They are too many files.

> @vyankyGH can we accept this in 1.4.3 now?

Yes I have tested with 1.4.3. Work's fine.

vyankyGH commented:
> @viveksahu26 see above

@viveksahu26 can you please test with current test command mutate changes.
**kyverno test https://github.com/sotoiwa/policies/patch-1**

viveksahu26 commented:
> 
> 
> > @viveksahu26 see above
> 
> @viveksahu26 can you please test with current test command mutate changes.
> **kyverno test https://github.com/sotoiwa/policies/patch-1**

All of them are fails, as it doesn't has kind field.
![image](https://user-images.githubusercontent.com/65091408/133726641-6b236304-9ea9-4841-958e-fbbc950d0128.png)


![image](https://user-images.githubusercontent.com/65091408/133726757-0839608f-af11-4bca-9467-7213c443b96e.png)

chipzoller commented:
@sotoiwa would you be able to add the `kind` field to each of the test cases, please? This is a new requirement as of 1.4.3.

sotoiwa commented:
@chipzoller I could rebased the branch with no conflicts. So all the tests should now have `kind` field.

viveksahu26 commented:
Hey @chipzoller , one question. Is `validationActionFailue: true`  for all validate policy.

chipzoller commented:
Most validate policies should be set to `audit` except for the ones that make no sense to be like Pod exec policies.
